README.txt


Ingres Python DBI Driver Version 2.0.1

-------------------------------------------------------------------------------

  1. Welcome

       1. New_in_This_Release

  2. Operating_System_Support
  3. Installation_Considerations
  4. General_Considerations

       1. Features_Not_Included
       2. Syntax_for_the_ingresdbi.connect()_Method
       3. Syntax_for_the_Ingres_Extension_Cursor.prepared_Attribute

  5. Building_and_Installing_the_Ingres_Python_DBI_Driver

       1. Building_the_Driver
       2. Installing_the_Driver

  6. Example_Code
  7. Known_Issues

       1. Generic_Issues
       2. Windows_Issues

  8. Support

-------------------------------------------------------------------------------

Welcome

This README contains all the documentation on the Ingres Python DBI driver.
Please review this README before building or installing this software. We
encourage users to test the software and provide feedback.
-------------------------------------------------------------------------------

New in This Release

The following changes have been made to the driver since its last release:

* Raise IOError if the trace file cannot be opened
* Extend search for odbcinst.ini

See the CHANGELOG for a complete listing of all changes.
-------------------------------------------------------------------------------

Operating System Support

This Ingres Python DBI driver supports all of the platforms supported by
Ingres, including:

* Solaris
* HP-UX
* AIX
* Linux
* Windows
* SCO UnixWare
* SCO OpenServer

-------------------------------------------------------------------------------

Installation Considerations

To build and install the Ingres Python DBI interface, the following components
are needed:

* Ingres r3 or above, including Ingres 2006. For a list of binary downloads,
  see http://www.ingres.com. For a source listing if you wish to build Ingres
  from source code, see http://www.ingres.com.
* C compiler (for example, GNU/C or Microsoft Visual Studio)
* Python interpreter version 2.4 or above
* The Ingres Python DBI source code

-------------------------------------------------------------------------------

General Considerations


Features Not Included

The following features are currently not included in the Ingres Open Source
Python DBI driver:

* Connection pooling (non-Windows only)
* The following extended Cursor attributes and methods:

  o messages
  o lastrowid
  o scroll

* Due to the limitations of the Ingres ODBC driver, the following items are not
  supported:

  o Executing functions asynchronously
  o Cursor direction other than forward-only
  o Support for Ingres SQL command "COPY TABLE"
  o Support for Ingres SQL command "SAVEPOINT"

* Due to syntax limitations of the Cursor.callproc() method, BYREF and output
  parameters are not supported in stored procedures. Row-returning procedures,
  however, are supported.

-------------------------------------------------------------------------------

Syntax for the ingresdbi.connect() Method

Connection objects are constructed using the ingresdbi.connect() method. The
following keywords are valid:

Keyword          Description
dsn              the ODBC Data Source Name
database         the target database
connectstr       alternate connection string
vnode            vnode definition as defined in the Ingres netutil utility. For
                 more information, see the Ingres_Connectivity_Guide
uid              user login ID
pwd              user password
servertype       the type of the target database
trace            enables optional tracing of the DBI driver
rolename         role name
rolepwd          role password
group            group
dbms_pwd         DBMS password
selectloops      fetches using select loops instead of the default cursor loops
autocommit       whether autocommit is enabled. Autocommit is off by default
catschemanull    whether to disable underscores in wildcard searches
catconnect       whether to use separate sessions for catalog operations
numeric_overflow whether to ignore or fail numeric overflows

If the "dsn" keyword is specified, the other keywords are optional. If the
database is specified, the other keywords are optional. If the vnode keyword is
specified, and the connection is local, the value "(LOCAL)" can be used as the
vnode definition, or the vnode attribute can be omitted.
If the "connectstr" keyword is specified, the other keywords are optional. The
"connectstr" keyword specifies an ODBC connection string. For examples of valid
Ingres ODBC connection strings, see the Ingres_Connectivity_Guide.
All of the above keywords reference string values except for the "trace"
keyword. The "trace" keyword references a tuple with two members. The first
member is the tracing level, which can be a value of 0 through 7. The second
member is a string that describes the trace file. If the second member has a
value of "None", the tracing is written to the standard output.
The following values are valid for the "autocommit", "selectloops",
"catconnect", "catschemanull", and "numeric_overflow" keyword attributes:

* "on"
* "off"
* "y"
* "n"
* "Yes"
* "No"

The following values are valid for the "servertype" keyword:

* "INGRES"
* "DCOM"
* "IDMS"
* "DB2"
* "IMS"
* "ODBC"
* "VSAM"
* "RDB"
* "STAR"
* "RMS"
* "ORACLE"
* "INFORMIX"
* "SYBASE"
* "MSSQL"
* "DB2UDB"

If "INGRES" is not specified, the "servertype" values require access to an
Ingres (that is, Enterprise Access) or EDBC gateway server. Otherwise, no
gateway is required. The default is "INGRES".
Select loops usually have the best performance. However, only one select loop
can be active at a time. Cursor loops support unlimited multiple active result
sets, but can be slower in performance.
Following is an example of a valid instantiation of the ingresdbi connection
object, using all keywords:

      conn = ingresdbi.connect(dsn ="myDSN",
              database  "myDB",
              vnode = "(LOCAL)",
              uid = "myUID",
              pwd = "myPWD",
              dbms_pwd = "myDbmsPWD",
              group = "myGroup",
              rolename = "myRoleName",
              rolepwd = "myRolePwd",
              selectloops = "Y",
              autocommit = "Y",
              servertype = "INGRES",
              driver = "Ingres",
              catschemanull = "off",
              catconnect = "Off",
              numeric_overflow = "yes",
              connectStr = "DSN=myDSN",
              trace = (7, "dbi.log")
              )

Connection objects can be constructed without keywords. If keywords are not
used, arguments must follow the order: dsn, database, vnode, uid, pwd,
selectloops, autocommit, servertype, and trace. An example without keywords is
shown here:

      conn = ingresdbi.connect(
              "myDSN",
              "myDB",
              "(LOCAL)",
              "MnyUID",
              "myPWD",
              "Y",
              "Yes",
              "INGRES",
              "Ingres 3.0",
              "myRoleName",
              "myrolePWD",
              "myGroup",
              "n",
              "NO",
              "YES",
              "yes",
              "N",
              "myDbmsPwd",
              "DSN=myDSN",
              (7,"dbi.log")
              )

-------------------------------------------------------------------------------

Syntax for the Ingres Extension Cursor.prepared Attribute

Although PEP 249 requires all queries to be prepared, the Ingres DBI driver
does not prepare queries by default. Instead, the Cursor attribute "prepared"
can be deployed.
If Cursor.prepared is set to "y", "yes", or "on", subsequent queries specified
in the Cursor.execute() or Cursor.executemany() methods are executed as
prepared. However, only one query string is allowed for each cursor
instantiation. If the query string is changed, a warning is issued, and the
Ingres DBI driver resorts to direct query execution.
The Cursor.callproc() method raises an exception if the Cursor.prepared
attribute is set to "y", "yes", or "on".
The following values (entered in either uppercase or lowercase) are valid for
the Cursor.prepared attribute:

* "on"
* "off"
* "y"
* "n"
* "Yes"
* "No"

-------------------------------------------------------------------------------

Building and Installing the Ingres Python DBI Driver


Building the Driver

The build process has been simplified by the use of the Python DistUtils
package. Start the build process by extracting the necessary files from the
Ingres DBI compressed archive:

Mac OS X, Linux and Unix


  1. Extract the files from ingresdbi-2.0.1.tar.gz:

           gzip -cd ingresdbi-2.0.1.tar.gz | tar zvf -

     or if you have GNU tar:

           tar zxvf ingresdbi-2.0.1.tar.gz

  2. Enter the newly created source directory, ingresdbi-2.0.1:

           cd ingresdbi-2.0.1

  3. Initiate the build process.

           python setup.py build



Windows


* Use WinZip (or similar product) to extract the directories and files from
  ingresdbi-2.0.1.zip.

  o Cygwin users can use either of the commands for Mac OS X, Linux and Unix

* Enter the newly created source directory, ingresdbi-2.0.1:

        cd ingresdbi-2.0.1

* Initiate the build process.

        python setup.py build


Note: You can skip this and jump straight to the install process as this will
automatically build.
-------------------------------------------------------------------------------

Installing the Driver

As with the build process, the installation process makes use of DistUtils. By
default, the Ingres Python DBI driver is installed into Python's site-packages
directory. The ability to provide alternate installation locations has not been
investigated at this time. The only requirement for installing is to be able to
write to the site-packages directory.
To install, execute the following command:

      python setup.py install

To create a deliverable source package, execute the following command:

      python setup.py sdist

-------------------------------------------------------------------------------

Example Code

The following code provides a simple Python database example using the Ingres
Python DBI driver:

      import ingresdbi
      import pprint
      """
      import os
      username=os.getenv('test_username')
      password=os.getenv('test_password')
      vnode=os.getenv('test_vnode')
      database=os.getenv('test_database')
      trace=None
      """
      enable_trace = 0
      if enable_trace == 1:
          trace=(7, None)

CHANGELOG
IngresDBI CHANGELOG - Last updated $Date$



SVN
      Fix Unicode string truncation [clach04 - 04-dec-2008]
      Fix strange string corruption for immutable python strings [grant - 10-jul-2008]
      Add librt to allow building with Ingres 2006 r3 [grant - 8-Jul-2008]
      Include GPL v2 License text in all source files - [grant - 8-May-2008]
      Raise IOError if the trace file cannot be opened - [grant - 28-Apr-2008]
      Extend search for odbcinst.ini - [grant - 28-Apr-2008]

09-May-2006 - 2.0.0
      Incremented version to 2.0.0.
      Fixed a number of memory leaks and negative reference counts.
      Added cursor.setinputsize(), cursor.setoutputsizes(),
      cursor.errorhandler, connection.errorhandler, connection exception
      objects, warnings for extended attributes, cursor.messages, and
      connection.messages.

10-Mar-2006 - 1.9.4
      Incorporated markdown utility for generation of README.html.  Fixed
      various memory leaks related to Curosr.close().

26-Jan-2006 - 1.9.3
      Fixed reference count in Cursor.next() method.
      
22-Nov-2005 - 1.9.2
      Implemented extended cursor methods __iter__() and next().
      Implemented extended cursor attribute rownumber.

16-Nov-2005 - 1.9.1
      Populated cursor description attributes display size, internal size,
      precision, and scale.  Updated README.html file.

24-Jun-2005 - 1.9.0
      Added cursor.callproc().  Added datetime support.  Note that the 2.4 
      version of Python is now required to support the Ingres DBI.

19-May-2005 - 1.5.0
      Implemented cursor.executemany().  Upgraded cursor.callproc() to 
      support row-returning procedures and to return the input parameters.

23-Feb-2005 - 1.0.8
      Implemented cursor.fetchmany().  First phase of support for
      cursor.callproc(): no support for BYREF or return parameters,
      but does support input parameters.     

16-Feb-2005 - 1.0.7
      Tightened up INCREF and DECREF's to improve memory efficiency.

31-Jan-2005 - 1.0.6
      Fixed fetch of long varchars in dbi_cursorFetchone().

10-Jan-2004 - 1.0.5
      Added dbi_freeData() routine to help resolve massive memory leaks.

21-Dec-2004 - 1.0.4
    - Added module descriptor attributes BINARY, NUMBER, STRING, and DATETIME.
      Added full support for "None" update parameters (Note: this requires an
      Ingres ODBC driver at level 3.50.00.46 or later).  
    - Added connection attribute "autocommit".

17-Nov-2004 - 1.0.3
    - Fixed random GPF occuring due to memory corruption

08-Nov-2004 - 1.0.2
    - Fixed data truncation of char/varchar types

05-Nov-2004 - 1.0.1
    - Fix memory leak in connection class

04-Nov-2004 - 1.0.0
    - Fetch blobs of type LONG VARCHAR into SQL_C_CHAR, otherwise data corruption happens.
    - Changed default driver to allow DSN less connections. removed references to ingres.lib preventing build with Windows.
    - Added support for BigInt
    - Added TRUE/FALSE defines for Linux
    - Made float4/8 support functional
    - Added version attribute to ingresdbi class.

23-Sept-2004 - 0.9.1
   - Fixed support for INTEGER1
   - Fixed miscellaneous problems that arose when performing 
     conformance testing including hangs and access violations.

20-Sept-2004
   - Initial Version


dbi/iidbiconn.c
/*
** Copyright (c) 2008 Ingres Corporation
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License version 2 as
** published by the Free Software Foundation.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License along
** with this program; if not, write to the Free Software Foundation, Inc.,
** 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
**
*/

# ifdef WIN32
# include <windows.h>
# endif
# include <sql.h>
# include <sqlext.h>
# include <iidbi.h>
# include <iidbiconn.h>
# include <iidbiutil.h>

# ifndef TRUE
# define TRUE 0
# endif

# ifndef FALSE
# define FALSE 0
# endif

/**
** Name: iidbiconn.c - Ingres Python DB API connection functions
**
** Description:
**     This file defines:
**(E
**   dbi_connect() Create a connection to the specified database.
**   dbi_connectionClose() Close the connection & issue a rollback.
**   dbi_connectionCommit() Commit the current transaction.
**   dbi_connectionCursor() Declare a cursor. Return the handle.
**   dbi_connectionRollback() Rollback the transaction.
** 		
**)E
**
** History:
**     08-jul-04 (peeje01)
**         Created.
**      10-Jul-2004 (raymond.fan@ca.com)
**          Add ODBC body to stubs.
**      10-Jul-2004 (raymond.fan@ca.com)
**          Moved print_err to iidbiutil.c.
**      10-Jul-2004 (raymond.fan@ca.com)
**          Pointer (self) from python was not declared to be pointer.
**          Corrected indirection error and replaced use of global IIDBIenv
**          with (self).
**      11-Jul-2004 (raymond.fan@ca.com)
**          Correct multiple connection handling.
**      11-Jul-2004 (raymond.fan@ca.com)
**          Add iidbutil.h for internal tracing.          
**      12-Jul-2004 (raymond.fan@ca.com)
**          Add return and tracing of SQL errors.
**      13-Jul-2004 (chris.clark@ca.com)
**          Centralize the determination of returned error status.
**      14-Jul-2004 (raymond.fan@ca.com)
**          Add pdbc references to the trace prints to show owners.
**      21-Oct-2004 (grant.croker@ca.com)
**          Corrected driver to call with an r3 installation
**      04-Nov-2004 (ralph.loen@ca.com)
**          In dbi_connectionClose(), free the connection handle
**          after disconnection.
**      21-Dec-2004 (ralph.loen@ca.com)
**          Added connection attribute "autocommit".        
**      15-dec-2005 (loera01)
**          Changed dbi_connect() so that only one argument is passed.
**          Add selectloops connection attribute.
**      26-jul-2005 (loera01)
**          Add servertype connection attribute.
**/

/*{
** Name: dbi_alloc_env
**
** Description:
**
**
** Inputs:
**     penv - DBI environment handle.
**     pdbc - DBI connectio handle.
**
** Outputs:
**     New ODBC environment handle.
**
** Returns:
**     SQL_SUCCESS
**     SQL_INVALID_HANDLE
**     SQL_ERROR
** 
** Side Effects:
**     Updates global DBI environment handle.
**
** History:
**      05-June-2006 (loera01)
**          Created.
}*/

RETCODE
dbi_alloc_env( IIDBI_ENV *penv, IIDBI_DBC *pdbc)
{
    HENV henv;
    RETCODE rc, return_code = DBI_SQL_SUCCESS;
    IIDBI_CONNECTION *conn = pdbc->conn;

    DBPRINTF(DBI_TRC_ENTRY)( "%p: dbi_alloc_env }}}1\n", pdbc );

    for (;;)
    {
        if (conn->pooled)
        {
            rc = SQLSetEnvAttr(SQL_NULL_HANDLE, SQL_ATTR_CONNECTION_POOLING,
                (SQLPOINTER)SQL_CP_ONE_PER_HENV,  sizeof(SQLINTEGER));
            if (rc != SQL_SUCCESS)
            {
                return_code = IIDBI_ERROR( rc, NULL, NULL, NULL, &pdbc->hdr.err );
                DBPRINTF(DBI_TRC_STAT)( "%p: %d = SQLSetEnvAttr (%d) %s %s %x\n", 
                    pdbc, rc, __LINE__, pdbc->hdr.err.sqlState, 
                    pdbc->hdr.err.messageText,
                    pdbc->hdr.err.native );
                break;
            }
        }
        rc = SQLAllocHandle(SQL_HANDLE_ENV, NULL, &henv);
        if (rc != SQL_SUCCESS)
        {
            return_code = IIDBI_ERROR( rc, henv, NULL, NULL, &pdbc->hdr.err );
                DBPRINTF(DBI_TRC_STAT)( "%p: %d = SQLAllocHandle SQL_HANDLE_ENV (%d) %s %s %x\n", pdbc,
                rc, __LINE__, pdbc->hdr.err.sqlState, pdbc->hdr.err.messageText,
                pdbc->hdr.err.native );
                break;
        }

        penv->hdr.handle = henv;

        rc = SQLSetEnvAttr(henv, SQL_ATTR_ODBC_VERSION,
            (void*)SQL_OV_ODBC3, 0);
        if (rc != SQL_SUCCESS)
        {
            return_code = IIDBI_ERROR( rc, henv, NULL, NULL, &pdbc->hdr.err );
            DBPRINTF(DBI_TRC_STAT)( "%p: %d = SQLSetEnvAttr (%d) %s %s %x\n", pdbc,
                rc, __LINE__, pdbc->hdr.err.sqlState, pdbc->hdr.err.messageText,
                pdbc->hdr.err.native );
            break;
        }
        break;
    }
    DBPRINTF(DBI_TRC_ENTRY)( "%p: dbi_alloc_env {{{1\n", pdbc );
    return return_code;
}

/*{
** Name: dbi_connect	-Create a connection to the specified database.
**
** Description:
**		Create a connection to the specified database.
**              The parameter dsn is the only required parameter, all others
**              are optional. The parameter list is DBMS dependent/defined(?)
**
** Assumptions:
**     The following global variables must be set:
**(E
**     Name           Meaning (possible values)
**     apilevel      -DB API Level ('1.0' or '2.0')
**     thread safety -Thread safety (0, 1, 2, 3)
**     paramstyle    -Parameter style ('qmark', 'numeric', 'format', 'pyformat')
**)E
**
** Inputs:
**     IIDBI_DBC pdbc Control block
**     char *dsn      Data Source Name
**     char *user     User name
**     char *password Password
**     char *host     Hostname
**     char *database Database
**
** Outputs:
**     Returns:
**         Database Connection
**     Exceptions:
**        Warning           E.G. for data truncations
**        InterfaceError    Database interface errors
**        DatabaseError     Database errors
**        OperationalError  Errors in database operation
**        IntegrityError    Relational Integrity failure
**        InternalError     Internal DB error
**        ProgrammingError  Nonexistant Table, syntax, ...
**        NotSupportedError Unsupported API method used
**        Error             All other errors
** 	
** Side Effects:
**
** History:
**      08-Jul-2004 (peeje01)
**          Created.
**      10-Jul-2004 (raymond.fan@ca.com)
**          Add ODBC connect code from clach04, loera01 
**      11-Jul-2004 (raymond.fan@ca.com)
**          Assign IIDBIenv with initial value on initialization.
**          Initialize henv from IIDBIenv.
}*/
RETCODE
dbi_connect( IIDBI_DBC *pdbc)
{
    RETCODE     rc = SQL_SUCCESS;
    HDBC        hdbc=NULL;
    HENV        henv=pdbc->env;
    IIDBI_CONNECTION *conn = pdbc->conn;
    char        szConnStrIn[300]= "\0";
    char        szConnStrOut[300] = "\0";
    SQLSMALLINT cbConnStrOut;
    char        *serverStr="SERVER=%s;";
    char        *dsnStr="DSN=%s;";
    char        *uidStr="UID=%s;";
    char        *pwdStr="PWD=%s;";
    char        *dbStr="DATABASE=%s;";
    char        *sloopStr="SELECTLOOPS=%s;";
    char        *stypeStr="SERVERTYPE=%s;";
    char        *driverStr="DRIVER=%s;";
    char        *rolenameStr="ROLENAME=%s;";
    char        *rolepwdStr="ROLEPWD=%s;";
    char        *groupStr="GROUP=%s;";
    char        *blankdateStr="BLANKDATE=%s;";
    char        *date1582Str="DATE1582=%s;";
    char        *catconnectStr="CATCONNECT=%s;";
    char        *numeric_overflowStr="NUMERIC_OVERFLOW=%s;";
    char        *catschemanullStr="CATSCHEMANULL=%s;";
    char        *dbms_pwdStr="DBMS_PWD=%s;";
    char        connStr[300];
    int         return_code = DBI_SQL_ERROR;

    DBPRINTF(DBI_TRC_ENTRY)( "%p: dbi_connect {{{1\n", pdbc );
    for(;;)
    {
        conn = (IIDBI_CONNECTION *)pdbc->conn;
        rc = SQLAllocHandle(SQL_HANDLE_DBC, henv, &hdbc);
        if (rc == SQL_INVALID_HANDLE)
        {
            DBPRINTF(DBI_TRC_STAT)( "%p: %d = SQLAllocHandle SQL_HANDLE_DBC (%d)\n", pdbc,
                rc, __LINE__ );
            return_code = DBI_INTERNAL_ERROR;
            (void) dbi_error_withtext( rc, NULL, NULL, NULL, &pdbc->hdr.err, "SQLAllocHandle - Invalid handle." ); 
            break;
        }
        if (rc != SQL_SUCCESS) 
        {
            return_code = IIDBI_ERROR(rc, NULL, hdbc, NULL, &pdbc->hdr.err );
            DBPRINTF(DBI_TRC_STAT)( "%p: %d = SQLDriverConnect (%d) %s %s %x\n", pdbc,
                rc, __LINE__, pdbc->hdr.err.sqlState, pdbc->hdr.err.messageText,
                pdbc->hdr.err.native );
            break;
        }
        
        if (conn->connectstr && *conn->connectstr)
        {
            strcpy(szConnStrIn, conn->connectstr);
        }
        else
        {
            if (conn->dsn && *conn->dsn)
            {
                sprintf(szConnStrIn, dsnStr, conn->dsn);
            }
            else if (!conn->driver)
            {
                sprintf(szConnStrIn, "%s", "DRIVER=INGRES 3.0;");
            }
            else if (conn->driver && *conn->driver)
            {
                sprintf(connStr, driverStr, conn->driver);
                strcat(szConnStrIn, connStr);
            }
            if (conn->vnode && *conn->vnode)
            {
                sprintf(connStr, serverStr, conn->vnode);
                strcat(szConnStrIn, connStr);
            }
            if (conn->database && *conn->database)
            {
                sprintf(connStr, dbStr, conn->database);
                strcat(szConnStrIn, connStr);
            }
            if (conn->uid && *conn->uid)
            {
                sprintf(connStr, uidStr, conn->uid);
                strcat(szConnStrIn, connStr);
            }
            if (conn->pwd && *conn->pwd)
            {
                sprintf(connStr, pwdStr, conn->pwd);
                strcat(szConnStrIn, connStr);
            }
            if (conn->selectloops)
            {
                sprintf(connStr, sloopStr, "Y");
                strcat(szConnStrIn, connStr);
            }
            else
            {
                sprintf(connStr, sloopStr, "N");
                strcat(szConnStrIn, connStr);
            }
            if (conn->blankdate)
            {
                sprintf(connStr, blankdateStr, "N");
                strcat(szConnStrIn, connStr);
            }
            if (conn->date1582)
            {
                sprintf(connStr, date1582Str, "N");
                strcat(szConnStrIn, connStr);
            }
            if (conn->catconnect)
            {
                sprintf(connStr, catconnectStr, "N");
                strcat(szConnStrIn, connStr);
            }
            if (conn->numeric_overflow)
            {
                sprintf(connStr, numeric_overflowStr, "N");
                strcat(szConnStrIn, connStr);
            }
            if (conn->catschemanull)
            {
                sprintf(connStr, catschemanullStr, "N");
                strcat(szConnStrIn, connStr);
            }
            if (conn->servertype && *conn->servertype)
            {
                sprintf(connStr, stypeStr, conn->servertype);
                strcat(szConnStrIn, connStr);
            }
            if (conn->rolename && *conn->rolename)
            {
                sprintf(connStr, rolenameStr, conn->rolename);
                strcat(szConnStrIn, connStr);
            }
            if (conn->rolepwd && *conn->rolepwd)
            {
                sprintf(connStr, rolepwdStr, conn->rolepwd);
                strcat(szConnStrIn, connStr);
            }
            if (conn->group && *conn->group)
            {
                sprintf(connStr, groupStr, conn->group);
                strcat(szConnStrIn, connStr);
            }
            if (conn->dbms_pwd && *conn->dbms_pwd)
            {
                sprintf(connStr, dbms_pwdStr, conn->dbms_pwd);
                strcat(szConnStrIn, connStr);
            }
        }
        DBPRINTF(DBI_TRC_STAT)( "%p: Connecting to %s (%d)\n", pdbc, szConnStrIn, __LINE__ );
        rc = SQLDriverConnect(hdbc, 0,
            (SQLCHAR *)szConnStrIn, SQL_NTS, (SQLCHAR *)szConnStrOut, 
            sizeof(szConnStrOut), &cbConnStrOut, 
            SQL_DRIVER_NOPROMPT);
        if (rc == SQL_INVALID_HANDLE)
        {
            DBPRINTF(DBI_TRC_STAT)( "%p: %d = SQLDriverConnect (%d)\n", pdbc, rc, __LINE__ );
            return_code = DBI_INTERNAL_ERROR;
            (void) dbi_error_withtext( rc, NULL, NULL, NULL, &pdbc->hdr.err, "SQLDriverConnect - Invalid handle." ); 
            break;
        }
        if (rc != SQL_SUCCESS) 
        {
            return_code = IIDBI_ERROR( rc, NULL, hdbc, NULL, &pdbc->hdr.err );
            DBPRINTF(DBI_TRC_STAT)( "%p: %d = SQLDriverConnect (%d) %s %s %x\n", pdbc,
                rc, __LINE__, pdbc->hdr.err.sqlState, pdbc->hdr.err.messageText,
                pdbc->hdr.err.native );
            break;
        }
        DBPRINTF(DBI_TRC_STAT)( "%p: Connected to %s (%d)\n", pdbc, szConnStrOut, __LINE__ );
        pdbc->hdr.handle = hdbc;
        if (conn->autocommit)
        {
             DBPRINTF(DBI_TRC_STAT)
                 ( "%p: Connected. Setting autocommit to ON (%d)\n", pdbc,
                 __LINE__ );
            rc = SQLSetConnectAttr(hdbc,SQL_ATTR_AUTOCOMMIT,
                (SQLPOINTER)SQL_AUTOCOMMIT_ON, sizeof(SQLINTEGER));
        }
        else
        {
            DBPRINTF(DBI_TRC_STAT)
                ( "%p: Connected. Setting autocommit to OFF (%d)\n", pdbc,
                __LINE__ );
            rc = SQLSetConnectAttr(hdbc,SQL_ATTR_AUTOCOMMIT,
                (SQLPOINTER)SQL_AUTOCOMMIT_OFF, sizeof(SQLINTEGER));
        }
        if (rc == SQL_INVALID_HANDLE)
        {
            DBPRINTF(DBI_TRC_STAT)
                ( "%p: %d = SQLSetConnectAttr SQL_ATTR_AUTOCOMMIT (%d)\n", pdbc,
                rc, __LINE__ );
            return_code = DBI_INTERNAL_ERROR;
            (void) dbi_error_withtext( rc, NULL, NULL, NULL, 
                &pdbc->hdr.err, "SQLSetConnectAttr - Invalid handle." ); 
            break;
        }
        if (rc != SQL_SUCCESS) 
        {
            return_code = IIDBI_ERROR( rc, NULL, hdbc, NULL, &pdbc->hdr.err );
            DBPRINTF(DBI_TRC_STAT)( "%p: %d = SQLConnectAttr SQL_AUTOCOMMIT_OFF (%d) %s %s %x\n", pdbc,
                rc, __LINE__, pdbc->hdr.err.sqlState, pdbc->hdr.err.messageText,
                pdbc->hdr.err.native );
            break;
        }
        return_code = DBI_SQL_SUCCESS;
        break;
    }
    DBPRINTF(DBI_TRC_ENTRY)("%p: dbi_connect }}}1\n", pdbc);
    return return_code;
}

/*{
** Name: dbi_connectionClose - Issue a rollback and close the connection.
**
Corrected driver to call with an r3 installatioCorrected driver to call with an r3 installation
** Inputs:
**     None.
**
** Outputs:
**     None.
**
** Side Effects:
**
** History:
**      08-Jul-2004 (peeje01)
**          Created.
**      10-Jul-2004 (raymond.fan@ca.com)
**          Add ODBC connect code from clach04, loera01 
}*/
RETCODE
dbi_connectionClose( IIDBI_DBC *pdbc )
{
    RETCODE rc;
    HDBC hdbc = pdbc->hdr.handle;
    int         return_code = DBI_SQL_SUCCESS;

    DBPRINTF(DBI_TRC_ENTRY)("%p: dbi_connectionClose {{{1\n", pdbc);
    
    if (hdbc)
    {
        rc = SQLDisconnect(hdbc);
        if (rc != SQL_SUCCESS) 
        {
            return_code = IIDBI_ERROR( rc, NULL, hdbc, NULL, &pdbc->hdr.err );
            DBPRINTF(DBI_TRC_STAT)( "%p: %d = SQLDisconnect (%d) %s %s %x\n", pdbc,
                rc, __LINE__, pdbc->hdr.err.sqlState, pdbc->hdr.err.messageText,
                pdbc->hdr.err.native );
        }
        else
        {
            DBPRINTF(DBI_TRC_STAT)( "%p: Disconnected successfully (%d)\n", pdbc, __LINE__ );
        }
    }

    rc = SQLFreeConnect( hdbc );
    if (rc == SQL_INVALID_HANDLE)
    {
        DBPRINTF(DBI_TRC_STAT)( "%p: Invalid ODBC connection handle (%d)\n", 
            pdbc, __LINE__ );
        if (return_code == DBI_SQL_SUCCESS)
            return_code = DBI_INTERNAL_ERROR;
    } 
    else if (rc == SQL_ERROR)
    {
        if (return_code == DBI_SQL_SUCCESS)
            return_code = IIDBI_ERROR( rc, NULL, NULL, 
                pdbc, &pdbc->hdr.err );
    
        DBPRINTF(DBI_TRC_STAT)( "%p: %d = dbi_connectionClose (%d) %s %s %x\n",
            pdbc, rc, __LINE__, pdbc->hdr.err.sqlState, 
            pdbc->hdr.err.messageText, pdbc->hdr.err.native );
    }

    DBPRINTF(DBI_TRC_ENTRY)("%p: dbi_connectionClose }}}1\n", pdbc);
    return return_code;
}

/*{
** Name: dbi_connectionCommit -Issue a commit
**
** Description:
**     Commit pending transaction.
**
** Inputs:
**     None.
**
** Outputs:
**     None.
**         Returns:
** 	       void
** Exceptions:
**     None.
**
** Side Effects:
**     None.
**
** History:
**     08-Jul-2004 (peeje01)
**         Created.
**      10-Jul-2004 (raymond.fan@ca.com)
**          Add ODBC connect code from clach04, loera01 
}*/
RETCODE
dbi_connectionCommit( IIDBI_DBC *pdbc )
{
    RETCODE rc;
    HDBC hdbc = pdbc->hdr.handle;
    int         return_code = DBI_SQL_ERROR;

    DBPRINTF(DBI_TRC_ENTRY)("%p: dbi_connectionCommit {{{1\n", pdbc);
    rc = SQLEndTran(SQL_HANDLE_DBC, hdbc, SQL_COMMIT);
    if (rc != SQL_SUCCESS) 
    {
        return_code = IIDBI_ERROR( rc, NULL, hdbc, NULL, &pdbc->hdr.err );
        DBPRINTF(DBI_TRC_STAT)( "%p: %d = SQLEndTran SQL_COMMIT (%d) %s %s %x\n", pdbc,
            rc, __LINE__, pdbc->hdr.err.sqlState, pdbc->hdr.err.messageText,
            pdbc->hdr.err.native );
    }
    else
    {
        DBPRINTF(DBI_TRC_STAT)( "%p: Committed successfully (%d)\n", pdbc, __LINE__ );
        return_code = DBI_SQL_SUCCESS;
    }
    DBPRINTF(DBI_TRC_ENTRY)("%p: dbi_connectionCommit }}}1\n", pdbc);
    return return_code;
}

/*{
** Name: dbi_connectionCursor - Return a cursor object
**
** Description:
**     Return a new cursor object using the connection.
**
** Inputs:
**     None.
**
** Outputs:
**     None.
**     Returns:
** 	   The new cursor object.
**     Exceptions:
**         None.
**
** Side Effects:
**
** History:
**     08-Jul-2004 (peeje01)
**         Created.
}*/
RETCODE
dbi_connectionCursor( IIDBI_DBC *pdbc )
{
    int         return_code = DBI_SQL_SUCCESS;
    DBPRINTF(DBI_TRC_ENTRY)("%p: dbi_connectionCursor {{{1\n", pdbc);
    DBPRINTF(DBI_TRC_ENTRY)("%p: dbi_connectionCursor }}}1\n", pdbc);
    return(return_code);
}

/*{
** Name: dbi_connectionRollback - Issue a rollabck
**
** Description:
**     Issue a rollback on the connection
**
** Inputs:
**     None.
**
** Outputs:
**     None.
**     Returns:
** 	   None.
**     Exceptions:
**         None.
**
** Side Effects:
**
** History:
**      08-Jul-2004 (peeje01)
**          Created.
**      10-Jul-2004 (raymond.fan@ca.com)
**          Corrected iiDBIenv to resolve.
}*/
RETCODE
dbi_connectionRollback( IIDBI_DBC *pdbc )
{
    RETCODE rc;
    HDBC hdbc = pdbc->hdr.handle;
    int         return_code = DBI_SQL_ERROR;
    
    DBPRINTF(DBI_TRC_ENTRY)("%p: dbi_connectionRollback {{{1\n", pdbc);

    rc = SQLEndTran(SQL_HANDLE_DBC, hdbc, SQL_ROLLBACK);
    if (rc != SQL_SUCCESS) 
    {
        return_code = IIDBI_ERROR( rc, NULL, hdbc, NULL, &pdbc->hdr.err );
        DBPRINTF(DBI_TRC_STAT)( "%p: %d = SQLEndTran SQL_ROLLBACK (%d) %s %s %x\n", pdbc,
            rc, __LINE__, pdbc->hdr.err.sqlState, pdbc->hdr.err.messageText,
            pdbc->hdr.err.native );
    }
    else
    {
        DBPRINTF(DBI_TRC_STAT)( "%p: Rolled back successfully (%d)\n", pdbc, __LINE__ );
        return_code = DBI_SQL_SUCCESS;
    }
    DBPRINTF(DBI_TRC_ENTRY)("%p: dbi_connectionRollback }}}1\n", pdbc);
    return return_code;
}
